Bus refactor

mprudnik opened the pull request:
Please, pay attention to bus only - `bus/bus.js` and `bus/types.d.ts`:
Events handling is still under work, mainly I was concentrated on the Command interface.
General flow:
1) Local call:
 - check if the service is registered in buss' `#localServices`
 - pass arguments to it
 - respond with the result
2) Remote call:
 - serialize call arguments
 - generate `callId`
 - create new event in `redis`'s stream with `serverId`, `callId` and serialized `payload`
 - get, deserialize and process it in consumer group
 - since `serverId` is known from metadata, respond through Pub/Sub

mprudnik committed:
bus - added initial implementation of bus over redis streams

mprudnik committed:
bus - updated few types and covered all bus methods with types

mprudnik committed:
general - updated lint config to parse d.ts files, fixed existing lint errors, updated infra folder structure

georgolden commented on src/infra/bus/bus.js:
```diff
@@ -0,0 +1,148 @@
+  #terminating = false;
+
+  /** @type init */
+  constructor({ redis }, { serverId }) {
```
IMO better to handle it as metadata

georgolden commented on src/infra/bus/bus.js:
```diff
@@ -0,0 +1,148 @@
+      serviceName,
+      this.#serverId,
+      streams,
+      { COUNT: 1, BLOCK: 10000, NOACK: true },
```
Make it `default = true` but has an option to enable acknowledgment mechanism?

georgolden reviewed (commented):
IMO it is better to use Strategy and go with local first - then remote as a default first strategy but have a flexibility to change strategy in future.
Useful links:
https://github.com/HowProgrammingWorks/Strategy
https://refactoring.guru/design-patterns/strategy
https://refactoring.guru/design-patterns/template-method

mprudnik committed:
updated lint and types, fixed ts/eslint errors

mprudnik committed:
bus - separated local and distributed bus implementations

mprudnik committed:
bus - added bus teardown

mprudnik committed:
validation - updated bus to have get/set schema methods

mprudnik committed:
validation - updated server to get schema from bus, instead of HTTPRoute, and to properly process it

mprudnik committed:
validation - updated services types and init to allow schema definition for commands

mprudnik committed:
bus - fixed bus methods dont bind to this

mprudnik committed:
general - moved to yarn, added json-schema-to-ts, updated lib/schema to support conversion to types

mprudnik committed:
validation - updated auth/ api and services to use new format of validation definition

mprudnik committed:
validation - updated account services to use new validation format

mprudnik committed:
validation - updated notification service types

georgolden commented on src/infra/bus/distributed.js:
```diff
@@ -0,0 +1,184 @@
+    this.#redis.publish(`response:${serverId}:${callId}`, message);
+  }
+
+  /** @type ICommand['call'] */
+  async #localCall({ service: serviceName, method }, payload) {
+    const service = this.#services.get(serviceName);
+    const handler = service[method];
+    if (!handler)
+      return [{ expected: false, message: 'Method not found' }, null];
+    const result = await handler(payload);
+    return result;
+  }
```
Why you need local call here?

mprudnik replied:
To handle receiving remote commands.
Basically, it's just a helper function to extract the "passing remote call data to local service" logic.
I think it will be easier to maintain buses since `LocalBus` has the same function.

georgolden reviewed (changes requested):
Only 2 things. No test for distributed bus implemented and local call with the distributed bus.

mprudnik committed:
Merge pull request #7 from mprudnik/validation-refactor

Validation refactor

mprudnik committed:
bus - added local bus tests

mprudnik committed:
general - updated lint config and fixed lint errors

mprudnik committed:
bus - updated distributed implementation to properly listen for calls and events

mprudnik committed:
general - fixed lint errors

mprudnik committed:
bus - added tests for distributed bus, fixed setup and teardown steps

mprudnik committed:
general - fixed lint errors, clean up console calls, updated types

mprudnik commented:
@georgolden Updated to use strategy.
Additional changes:
 - updated distributed bus to initialize and teardown properly
 - added tests for both buses' implementation
 - added logging for distributed bus flows
 
 Missing `call/event tracing` and `error handling` to be ready for the full review.

mprudnik committed:
bus - added meta propagation and error handling for services and bus, improved types

mprudnik committed:
general - fixed lint errors, removed unnecessary lib/error

georgolden commented on src/infra/bus/distributed.js:
```diff
@@ -0,0 +1,378 @@
+
+  /** @type IBus['withMeta'] */
+  withMeta = (meta) => {
+    const allowedMethods = ['call', 'publish'];
```
unnecessary allocation, better use static field

georgolden commented on src/infra/bus/distributed.js:
```diff
@@ -0,0 +1,378 @@
+    const allowedMethods = ['call', 'publish'];
+    return new Proxy(this, {
+      get(target, prop) {
+        if (typeof prop === 'string' && allowedMethods.includes(prop)) {
+          return undefined;
+        }
```
this if statement must be reversed. call and publish are both type `string` and in `allowedMethods` but undefined returns!!!
```js
if (typeof prop !== 'string' || !allowedMethods.includes(prop)) ...
```

georgolden commented on src/infra/bus/local.js:
```diff
@@ -0,0 +1,100 @@
+    return true;
+  };
+
+  /** @type IBus['withMeta'] */
+  withMeta = (meta) => {
+    const allowedMethods = ['call', 'publish'];
+    return new Proxy(this, {
+      get(target, prop) {
+        if (typeof prop === 'string' && allowedMethods.includes(prop)) {
+          return undefined;
+        }
+
+        const method = target[prop];
+
+        return ({ meta: original, data }) =>
+          method({
+            data,
+            meta: { ...original, ...meta },
+          });
+      },
+    });
+  };
```
same issues as in DistributedBus implementation

georgolden commented on src/server/server.js:
```diff
@@ -22,19 +22,28 @@ export const init = async (infra, api, options) => {
-      ),
+        { meta: {}, data: { token, definition } },
+      );
+      if (error) {
+        if (error.expected)
```
```js
if (error?.expected) ...
```

mprudnik committed:
bus - fixed invalid withMeta condition and removed unnecessary allocation

mprudnik committed:
general - cleaned up verifyToken function of customAuth plugin

mprudnik commented:
@georgolden fixed
